pandas/core/indexes/datetimelike.py

```diff
@@ -662,7 +663,21 @@ def __add__(self, other):
                 return self._add_delta(other)
             elif is_integer(other):
                 return self.shift(other)
-            elif isinstance(other, (Index, datetime, np.datetime64)):
+            elif isinstance(other, (datetime, np.datetime64)):
+                return self._add_datelike(other)
+            elif (isinstance(self, DatetimeIndex) and is_offsetlike(other) and
+                  not isinstance(other, ABCSeries)):
+                # Array of DateOffset objects
+                if len(other) == 1:
+                    return self + other[0]
+                else:
+                    from pandas.errors import PerformanceWarning
```

Comment:
so you are adding code here that already exists in ``datetimes.py:_add_offset`` ? is there a reason you are not dispatching to that (which may need to handle array-likes in addition to scalars). Further I am not in love with explicit type checking for DTI here. again why are you not handling that at a higher level.

Reply:
_add_offset is for a single offset which does not have an implementation of `apply_index`.  This is for an array of offsets.  It's a case handled by Series but not by DatetimeIndex.

I can make a _add_offset_array and dispatch to that.